src/System.Net.Security/src/System/Net/Security/Sslstream.Implementation.cs

```diff
@@ -1473,6 +1199,414 @@ private void FinishHandshake(Exception e, AsyncProtocolRequest asyncRequest)
+        {
+            // Request a write IO slot.
+            Task ioSlot = writeAdapter.LockAsync();
+            if (!ioSlot.IsCompletedSuccessfully)
+            {
+                // Operation is async and has been queued, return.
+                return new ValueTask(WaitForWriteIOSlot(writeAdapter, ioSlot, buffer));
+            }
+
+            byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(buffer.Length + FrameOverhead);
+            byte[] outBuffer = rentedBuffer;
+
+            SecurityStatusPal status = EncryptData(buffer, ref outBuffer, out int encryptedBytes);
+
+            if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+            {
+                // Re-handshake status is not supported.
+                ArrayPool<byte>.Shared.Return(rentedBuffer);
+                ProtocolToken message = new ProtocolToken(null, status);
+                return new ValueTask(Task.FromException(new IOException(SR.net_io_encrypt, message.GetException())));
+            }
+
+            ValueTask t = writeAdapter.WriteAsync(outBuffer, 0, encryptedBytes);
+            if (t.IsCompletedSuccessfully)
+            {
+                ArrayPool<byte>.Shared.Return(rentedBuffer);
+                FinishWrite();
+                return t;
+            }
+            else
+            {
+                return new ValueTask(CompleteAsync(t, rentedBuffer));
+            }
+
+            async Task WaitForWriteIOSlot(TWriteAdapter wAdapter, Task lockTask, ReadOnlyMemory<byte> buff)
+            {
+                await lockTask.ConfigureAwait(false);
+                await WriteSingleChunk(wAdapter, buff).ConfigureAwait(false);
+            }
+
+            async Task CompleteAsync(ValueTask writeTask, byte[] bufferToReturn)
+            {
+                try
+                {
+                    await writeTask.ConfigureAwait(false);
+                }
+                finally
+                {
+                    ArrayPool<byte>.Shared.Return(bufferToReturn);
+                    FinishWrite();
+                }
+            }
+        }
+
+        //
+        // Validates user parameters for all Read/Write methods.
+        //
+        private void ValidateParameters(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (offset < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            if (count > buffer.Length - offset)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count), SR.net_offset_plus_count);
+            }
+        }
+
+        ~SslStream()
+        {
+            Dispose(disposing: false);
+        }
+
+        private void InternalDispose()
```

Comment:
Is this used anywhere other than CloseInternal?  If not, why is it separated out?

Reply:
removed